[nrunner] Add support to variants on nrunner

willianrampazzo opened the pull request:
This PR adds support to variants on nrunner. Tests are still missing.

To test it you can use one of the available yaml files, like:

```
[wrampazz@wrampazz avocado.dev]$ avocado run --test-runner=nrunner -m examples/tests/sleeptest.py.data/sleeptest.yaml -- examples/tests/sleeptest.py
JOB ID     : 264b50b21a505011bf38aafac58f7cb40d73780f
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-08-05T16.51-264b50b/job.log
 (1/4) examples/tests/sleeptest.py:SleepTest.test;run-short-beaf: STARTED
 (3/4) examples/tests/sleeptest.py:SleepTest.test;run-long-f397: STARTED
 (2/4) examples/tests/sleeptest.py:SleepTest.test;run-medium-5595: STARTED
 (4/4) examples/tests/sleeptest.py:SleepTest.test;run-longest-efc4: STARTED
 (1/4) examples/tests/sleeptest.py:SleepTest.test;run-short-beaf: PASS (0.51 s)
 (2/4) examples/tests/sleeptest.py:SleepTest.test;run-medium-5595: PASS (1.01 s)
 (3/4) examples/tests/sleeptest.py:SleepTest.test;run-long-f397: PASS (5.02 s)
 (4/4) examples/tests/sleeptest.py:SleepTest.test;run-longest-efc4: PASS (10.03 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-08-05T16.51-264b50b/results.html
JOB TIME   : 12.01 s
[wrampazz@wrampazz avocado.dev]$
```

beraldoleal commented on avocado/plugins/runner_nrunner.py:
```diff
@@ -188,7 +190,12 @@ def _create_runtime_tasks_for_test(test_suite, runnable, no_digits,
         # test related operations
+        # when creating variants, they need to have a copy of the runnable.
+        if len(test_suite.tests) == 1 and index > 1:
+            runnable = deepcopy(runnable)
```
:+1: ;)

willianrampazzo replied:
Hehe, thanks for the insight!

beraldoleal reviewed (approved):
LGTM, thanks.

willianrampazzo commented:
@clebergnu @beraldoleal, FYI, while I was investigating #4840, I found a problem in this code: 

```
[wrampazz@wrampazz avocado.dev]$ avocado-runner-avocado-instrumented runnable-run -u /tmp/foo.py:First.test_pass
{'status': 'started', 'time': 64065.782545549}
{'type': 'stderr', 'log': b'Traceback (most recent call last):\n  File "/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n', 'encoding': 'utf-8', 'status': 'running', 'time': 64065.786519159}
{'result': 'error', 'fail_reason': "'NoneType' object is not subscriptable", 'status': 'finished', 'time': 64065.786647994}
```

I changed the logic inside `avocado/core/runners/avocado_instrumented.py`, the last commit, to avoid the problem.

beraldoleal commented:
> @clebergnu @beraldoleal, FYI, while I was investigating #4840, I found a problem in this code:
> 
> ```
> [wrampazz@wrampazz avocado.dev]$ avocado-runner-avocado-instrumented runnable-run -u /tmp/foo.py:First.test_pass
> {'status': 'started', 'time': 64065.782545549}
> {'type': 'stderr', 'log': b'Traceback (most recent call last):\n  File "/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n', 'encoding': 'utf-8', 'status': 'running', 'time': 64065.786519159}
> {'result': 'error', 'fail_reason': "'NoneType' object is not subscriptable", 'status': 'finished', 'time': 64065.786647994}
> ```
> 
> I changed the logic inside `avocado/core/runners/avocado_instrumented.py`, the last commit, to avoid the problem.

Not sure if I understood, but the error on #4840 seems to be different. Using this patch the problem there is still there. I'm debugging there and it is related to the test path resolution.

richtja commented on avocado/core/runners/avocado_instrumented.py:
```diff
@@ -39,17 +40,28 @@ def _run_avocado(runnable, queue):
-            # this handles the -p command-line argument
-            params = TreeNode().get_node("/", True)
-            params.value = dict(runnable.config.get('run.test_parameters', []))
+            # inject the -p command-line / run.test_parameters option
```
I think it will be better to move parameters creation into its own method, for better readability.

willianrampazzo replied:
> > @clebergnu @beraldoleal, FYI, while I was investigating #4840, I found a problem in this code:
> > ```
> > [wrampazz@wrampazz avocado.dev]$ avocado-runner-avocado-instrumented runnable-run -u /tmp/foo.py:First.test_pass
> > {'status': 'started', 'time': 64065.782545549}
> > {'type': 'stderr', 'log': b'Traceback (most recent call last):\n  File "/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n', 'encoding': 'utf-8', 'status': 'running', 'time': 64065.786519159}
> > {'result': 'error', 'fail_reason': "'NoneType' object is not subscriptable", 'status': 'finished', 'time': 64065.786647994}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I changed the logic inside `avocado/core/runners/avocado_instrumented.py`, the last commit, to avoid the problem.
> 
> Not sure if I understood, but the error on #4840 seems to be different. Using this patch the problem there is still there. I'm debugging there and it is related to the test path resolution.

The error of #4840 still exists. What I meant was when I was debugging #4840, I was on the branch of this PR and it gave me the following error, which is related to the code of this PR:

```
"/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n'
```

I changed the logic to avoid the problem above.

willianrampazzo replied:
Do you mean to create a new method for the following 3 lines returning the `paths` and the `tree_nodes`? If that is the case, it is also worth adding a method to create the variant tree. I can do that, but I don't know if it will improve that much the readability. Let me know.

richtja replied:
No, I meant the whole part for creating `tree_nodes` and `path` the parameters part and also creating the variant tree. lines 43-57.

willianrampazzo replied:
Okay, got it. I'll change that.

willianrampazzo replied:
Done @richtja! Thanks for the review.

willianrampazzo commented:
> > @clebergnu @beraldoleal, FYI, while I was investigating #4840, I found a problem in this code:
> > ```
> > [wrampazz@wrampazz avocado.dev]$ avocado-runner-avocado-instrumented runnable-run -u /tmp/foo.py:First.test_pass
> > {'status': 'started', 'time': 64065.782545549}
> > {'type': 'stderr', 'log': b'Traceback (most recent call last):\n  File "/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n', 'encoding': 'utf-8', 'status': 'running', 'time': 64065.786519159}
> > {'result': 'error', 'fail_reason': "'NoneType' object is not subscriptable", 'status': 'finished', 'time': 64065.786647994}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I changed the logic inside `avocado/core/runners/avocado_instrumented.py`, the last commit, to avoid the problem.
> 
> Not sure if I understood, but the error on #4840 seems to be different. Using this patch the problem there is still there. I'm debugging there and it is related to the test path resolution.

The error of #4840 still exists. What I meant was when I was debugging #4840, I was on the branch of this PR and it gave me the following error, which is related to the code of this PR:

```
"/home/wrampazz/src/avocado/avocado.dev/avocado/core/runners/avocado_instrumented.py", line 45, in _run_avocado\n    for path, env in runnable.variant[\'variant\']]\nTypeError: \'NoneType\' object is not subscriptable\n'
```

I changed the logic to avoid the problem above.

willianrampazzo commented:
> @willianrampazzo this greatly advances the state of the varianter support for the nrunner, thanks! I'm betting you're planning to add support to other test types, right? The one I miss the most, and I'd like to use ASAP is `exec-test`.
> 
> Either way, I see no problem in including this version and working on top of this.

Thanks for the review, I'll work on other test types on top of this PR. TBH, I forgot about those :D
